security/security-api/custom-identity-store/README.md

```diff
@@ -0,0 +1,175 @@
+# Custom Identity Store
+
+An application also can provide its own IdentityStore. Bundled with the application, this is the identity store, which is then used for authentication, authorization.
+
+This example demonstrates how a custom identity store `TestIdentityStore` can be defined and provided as part of the application being deployed.
+
+In this example, following user is defined along with the roles they are in.
+
+|User|Password|Role|
+|----|--------|----|
+|reza|secret1|foo,bar|
+
+When a request is made to the application with certain credentials, the authentication mechanism bundled with this application comes into affect and an authentication is done against the `TestIdentityStore` as defined in the application.
+
+Post authentication, the application will also verify the roles the caller is in and will send the details as part of the response.
+
+Defining credentials and the roles assigned to the users is done as shown below in `TestIdentityStore`:
+
+```java
+if (usernamePasswordCredential.compareTo("reza", "secret1")) {
+    return new CredentialValidationResult("reza", new HashSet<>(asList("foo", "bar")));
+}
+
+```
+
+With application, a custom `HttpAuthenticationMechanism` too is bundled which gets called during authentication process.
+
+```java
+@RequestScoped
+public class TestAuthenticationMechanism implements HttpAuthenticationMechanism {
+
+    @Inject
+    private IdentityStoreHandler identityStoreHandler;
+
+    @Override
+    public AuthenticationStatus validateRequest(HttpServletRequest request, HttpServletResponse response, HttpMessageContext httpMessageContext) throws AuthenticationException {
+
+        if (request.getParameter("name") != null && request.getParameter("password") != null) {
+
+            // Get the (caller) name and password from the request
+            // NOTE: This is for the smallest possible example only. In practice
+            // putting the password in a request query parameter is highly
+            // insecure
+            String name = request.getParameter("name");
+            Password password = new Password(request.getParameter("password"));
+
+            // Delegate the {credentials in -> identity data out} function to
+            // the Identity Store
+            // Password internally gets verified using the PasswordHash specified while defining DatabaseIdentityStore.
+            CredentialValidationResult result = identityStoreHandler.validate(
+                new UsernamePasswordCredential(name, password));
+
+            if (result.getStatus() == VALID) {
+                // Communicate the details of the authenticated user to the
+                // container. In many cases the underlying handler will just store the details
+                // and the container will actually handle the login after we return from
+                // this method.
```

Comment:
... In many cases, the underlying handler stores the details and the container handles the login after this method is returned. (is that still technically correct?)
(same comment as previous example.)